### ISSUE TITLE
Fails to resolve `exports` defined submodule when reexported by another package

### ISSUE THREAD

**Issue by isaacs** [DataDog/import-in-the-middle#63]
`node_modules/has-submodule/package.json`
```json
{
  "name": "has-submodule",
  "type": "module",
  "exports": {
    "./sub": "./sub.js"
  }
}
```

`node_modules/has-submodule/sub.js`
```js
export const foo = 'bar'
console.log('loaded submodule')
cat: node_modules/reexport-submodule/sub.js: No such file or directory
```

`node_modules/reexport-submodule/package.json`
```json
{
  "name": "reexport-submodule",
  "exports": {
    ".": "./index.js"
  },
  "type": "module"
}
```

`node_modules/reexport-submodule/index.js`
```js
export * from 'has-submodule/sub'
```

`load-sub.mjs`

```js
import { foo } from 'reexport-submodule'
console.log(foo)
```

`esm-loader.mjs`

```js
export * from 'import-in-the-middle/hook.mjs'
```

## Expected Behavior

```
$ node load-sub.mjs
loaded submodule
bar
```

```
$ node --loader=./esm-loader.mjs load-sub.mjs
loaded submodule
bar
```

## Actual Behavior

```
$ node --loader=./esm-loader.mjs load-sub.mjs
(node:36249) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("./esm-loader.mjs", pathToFileURL("./"));'
(Use `node --trace-warnings ...` to show where the warning was created)

node:internal/process/esm_loader:40
      internalBinding('errors').triggerUncaughtException(
                                ^
[Error: ENOENT: no such file or directory, open '/Users/isaacs/dev/tapjs/esm-tap-repro/node_modules/reexport-submodule/has-submodule/sub'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/isaacs/dev/tapjs/esm-tap-repro/node_modules/reexport-submodule/has-submodule/sub'
}

Node.js v20.9.0
```

## Steps to Reproduce the Problem

Shown above.

**Comment by jsumners-nr** [comment]
At https://github.com/DataDog/import-in-the-middle/blob/c3c2c52c1915b47994af59d507c59029c1f1fae9/hook.js#L156-L157

When `modFile` is `has-submodule/sub`, then `modUrl` is defined as `file:///private/tmp/29/tap-issue/node_modules/reexport-submodule/has-submodule/sub`.

**Comment by meyer9** [comment]
Similar issue trying to import langchain:

```
modUrl: file:///Users/meyer9/development/repro/node_modules/langchain/dist/embeddings/@langchain/core/embeddings
modFile: @langchain/core/embeddings
normalizedModName: embeddings
srcUrl: file:///Users/meyer9/development/repro/node_modules/langchain/dist/embeddings/base.js
```

```
node:internal/process/esm_loader:40
      internalBinding('errors').triggerUncaughtException(
                                ^
[Error: ENOENT: no such file or directory, open '/Users/meyer9/development/repro/node_modules/langchain/dist/embeddings/@langchain/core/embeddings'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/meyer9/development/repro/node_modules/langchain/dist/embeddings/@langchain/core/embeddings'
}
```

The line in question:

```
export * from "@langchain/core/embeddings";
```

**Comment by ida613** [comment]
Not able to reproduce this issue, can you confirm the IITM version you are using?

**Comment by khanayan123** [comment]
I was able to reproduce the issue on master, looking into it

**Comment by chentsulin** [comment]
@khanayan123 Is there any update? I'd like to help if there are some failed tests can be investigated.

**Comment by timfish** [comment]
I've been looking into this with the following reliable reproduction using Node v22.2.0:
```ts
import { register } from 'node:module';

register('import-in-the-middle/hook.mjs', import.meta.url);
await import ('@discordjs/builders');
```
Which results in:
```
node:internal/modules/run_main:125
    triggerUncaughtException(
    ^
Error: ENOENT: no such file or directory, open '/Users/tim/Documents/Repositories/repro/node_modules/@discordjs/builders/dist/@discordjs/formatters'
    at async open (node:internal/fs/promises:638:25)
    at async readFile (node:internal/fs/promises:1251:14)
    at async getSource (node:internal/modules/esm/load:46:14)
    at async defaultLoad (node:internal/modules/esm/load:137:34)
    at async nextLoad (node:internal/modules/esm/hooks:750:22)
    at async getExports (/Users/tim/Documents/Repositories/repro/node_modules/import-in-the-middle/lib/get-exports.js:64:21)
    at async processModule (/Users/tim/Documents/Repositories/repro/node_modules/import-in-the-middle/hook.js:136:23)
    at async processModule (/Users/tim/Documents/Repositories/repro/node_modules/import-in-the-middle/hook.js:164:20)
    at async getSource (/Users/tim/Documents/Repositories/repro/node_modules/import-in-the-middle/hook.js:274:60)
    at async load (/Users/tim/Documents/Repositories/repro/node_modules/import-in-the-middle/hook.js:340:26) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/tim/Documents/Repositories/repro/node_modules/@discordjs/builders/dist/@discordjs/formatters'
}
```

It appears that this code expects the module it's loading to be relative to the current file:
https://github.com/DataDog/import-in-the-middle/blob/00b01fff1f5b69dd25e307593ec54d1d8abb4844/hook.js#L154-L166

My proposal is that if the file cannot be found at the originally calculated relative URL, `resolve` is called to resolve the package name. Using resolve is better than manually trying to find the correct files since there might be multiple levels of `node_modules` directories that need traversing.

Initially I was hacking around attempting to fix this without a release of `import-in-the-middle` which resulted in this awful code which replaces `import-in-the-middle/hook.mjs` via another loader hook 😂:
```ts
import { register } from "node:module";

const overrideCode = `
import { stat } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { createHook } from "./hook.js";

const { load: loadIITM, resolve: resolveIITM, getFormat, getSource } = createHook(import.meta);

function getModuleName(url) {
    let parts = url.split("/@");
    if (parts.length > 1) {
        const last = parts.pop();
        if ((last.match("/") || []).length === 1) {
            return "@" + last;
        }
    } 
    return url.split("/").pop()
}

let parentResolve;

export function resolve(specifier, context, nextResolve = parentResolve) {
    parentResolve = nextResolve;
    return resolveIITM(specifier, context, nextResolve);
}

export async function load(url, context, nextLoad) {
    return loadIITM(url, context, async (u, c) => {
        const initialResult = await stat(fileURLToPath(u)).catch((e) => null);

        if (initialResult) 
          return nextLoad(u, c);

        const mod = getModuleName(u);
        // should probably set parentURL here
        const result = await resolve(mod, {});
        return nextLoad(result.url, c);
    });
}

export { getFormat, getSource };`;

// This loader overrides the source code of 'import-in-the-middle/hook.mjs' with the code above.
register(
  new URL(`data:application/javascript,
    export async function load(url, context, nextLoad) {
        if (url.endsWith('import-in-the-middle/hook.mjs')) {
            const result = await nextLoad(url, context);
            return {...result, source: '${overrideCode}' };
        }
        return nextLoad(url, context);
    }`
  ),
  import.meta.url
);

register("import-in-the-middle/hook.mjs", import.meta.url);

await import("@discordjs/builders");
```

This results in `SyntaxError: Identifier '$version' has already been declared` which is a completely different bug.

**Comment by timfish** [comment]
This should have been resolved by #78
